src/Microsoft.AspNetCore.Razor.Language/Legacy/MarkupParser.cs

```diff
@@ -0,0 +1,916 @@
+                    attributeBlock.Value.Children.Count > 0 &&
+                    IsTypeAttribute(attributeBlock))
+                {
+                    typeAttribute = attributeBlock;
+                    break;
+                }
+            }
+
+            if (typeAttribute != null)
+            {
+                var contentValues = typeAttribute.Value.CreateRed().DescendantNodes().Where(n => n.IsToken).Cast<Syntax.SyntaxToken>();
+
+                var scriptType = string.Concat(contentValues.Select(t => t.Content)).Trim();
+
+                // Does not allow charset parameter (or any other parameters).
+                return string.Equals(scriptType, "text/html", StringComparison.OrdinalIgnoreCase);
+            }
+
+            return false;
+        }
+
+        private static bool IsTypeAttribute(MarkupAttributeBlockSyntax attributeBlock)
+        {
+            if (attributeBlock.Name.LiteralTokens.Count == 0)
+            {
+                return false;
+            }
+
+            var trimmedStartContent = attributeBlock.Name.ToFullString().TrimStart();
+            if (trimmedStartContent.StartsWith("type", StringComparison.OrdinalIgnoreCase) &&
+                (trimmedStartContent.Length == 4 ||
+                ValidAfterTypeAttributeNameCharacters.Contains(trimmedStartContent[4])))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        // Internal for testing.
+        internal SyntaxToken AcceptAllButLastDoubleHyphens()
+        {
+            var lastDoubleHyphen = CurrentToken;
+            AcceptWhile(s =>
+            {
+                if (NextIs(SyntaxKind.DoubleHyphen))
+                {
+                    lastDoubleHyphen = s;
+                    return true;
+                }
+
+                return false;
+            });
+
+            NextToken();
+
+            if (At(SyntaxKind.Text) && IsHyphen(CurrentToken))
+            {
+                // Doing this here to maintain the order of tokens
+                if (!NextIs(SyntaxKind.CloseAngle))
+                {
+                    Accept(lastDoubleHyphen);
+                    lastDoubleHyphen = null;
+                }
+
+                AcceptAndMoveNext();
+            }
+
+            return lastDoubleHyphen;
+        }
+
+        private bool AcceptTokenUntilAll(in SyntaxListBuilder<RazorSyntaxNode> builder, params SyntaxKind[] endSequence)
+        {
+            while (!EndOfFile)
+            {
+                ParseMarkupNodes(builder, ParseMode.Text, t => t.Kind == endSequence[0]);
+                if (AcceptAll(endSequence))
+                {
+                    return true;
+                }
+            }
+            Debug.Assert(EndOfFile);
+            SpanContext.EditHandler.AcceptedCharacters = AcceptedCharactersInternal.Any;
+            return false;
+        }
+
+        private ParserState GetParserState(ParseMode mode)
```

Comment:
I'm a little skeptical about ParserMode's (instead of having a property that lives on the context of some sort) but we can see how this fleshes out once more things are added.

Reply:
I'm on the same page. I'll do what makes sense as I add more stuff.